### DISCUSSION TITLE
Error astype setelah replace string

### DISCUSSION THREAD

**Discussion by AngelaSherina** [teamalgoritma/community#296] [Q&A] [answered]
halo, saya ingin melakukan astype pada data clients setelah melakukan replace string namun terjadi error
mohon bantuan solving error nya. Terima Kasih

![Web capture_9-9-2022_154459_localhost](https://user-images.githubusercontent.com/112871237/189313746-a70b4967-8c5d-42b6-a85c-256050894e04.jpeg)
![Web capture_9-9-2022_154550_localhost](https://user-images.githubusercontent.com/112871237/189313761-1575c2c1-9629-49a6-a2ec-777d5c1648cc.jpeg)
![Web capture_9-9-2022_155016_localhost](https://user-images.githubusercontent.com/112871237/189313769-90d96591-d9f9-4f13-9a91-3171d917677f.jpeg)

**Comment by taruma** [comment, marked as answer]
Solusinya:

Terdapat koreksi pada _cell_ `In [7]` dan `In [9]`

`>>> In [7]`
```python
clients['Consulting Sales'] = clients['Consulting Sales'].str.replace('IDR', '')
```

`>>> In [9]`
```python
clients['Forecasted Growth'] = clients['Forecasted Growth'].str.replace('%', '')
```

dan setelah itu coba _execute_ kembali `In [18]`. Seharusnya proses _casting_ dari _string_ ke _integer_ berhasil. 

---

Jika membaca pesan error yang dilampirkan, dinyatakan bahwa ada hal yang _invalid_ saat proses _casting_ _object_ (`string`) ke _integer_ `int64` dengan _value_ yang gagal diubah yaitu `'IDR7125000'`. Jika kita coba melakukan perintah `int('IDR7125000')`, maka akan muncul pesan error yang serupa dengan _method_ `.astype(...)`.  

![image](https://user-images.githubusercontent.com/1007910/189336591-f9f059b5-356d-40a0-8da9-5425720961db.png)

Akan tetapi jika melakukan `int('7125000')`, tidak ada pesan error dan proses _casting_ berhasil. Setelah memahami kegagalan tersebut, dilanjutkan dengan pertanyaan "Kenapa replace tersebut tidak terganti?"

Error tersebut terjadi dikarenakan perubahan berupa `.str.replace(...)` tidak di _assign_ ke kolom yang seharusnya diubah. Ketik menggunakan _method_ dari _object_ `DataFrame/Series`, _object_ tersebut tidak dimodifikasi, dan hasil / _return_ dari _method_ dieksekusi tapi tidak direferensikan kemanapun. 

Ada beberapa _method_ yang menerima _argument_ `inplace=` untuk memberitahu "modifikasi langsung di _object_-nya tersebut". Akan tetapi, hal tersebut tidak direkomendasikan seperti yang diinformasikan di pandas-dev/pandas#16529. 

Jadi solusinya adalah selalu ingat untuk _assign_ hasil dari _method_ _object_ `DataFrame` ke _value_ yang memang perubahan tersebut dieksekusi. Kurang lebih umumnya seperti ini. 

```python
dataframe['NAMA_KOLOM'] = dataframe['NAMA_KOLOM'].method('ARGUMEN_METHOD')
```

Saya belum coba langsung di pythonnya, jadi mungkin bisa salah solusi. 🤣🤣. Tapi, semoga informasi yang saya tulis bisa bermanfaat di kemudian hari.

---

Bonus: karena ini saya kebanyakan nulis _assign_. di pandas sendiri ada _method_ bernama `.assign(...)`. Biar bisa lebih mudah dibaca, baris ini:

```python
dataframe['NAMA_KOLOM'] = dataframe['NAMA_KOLOM'].method('ARGUMEN_METHOD')
```

bisa diubah menjadi:

```python
dataframe = dataframe.assign(
    NAMA_KOLOM=dataframe['NAMA_KOLOM'].method('ARGUMEN_METHOD')
)
```

kapan digunakan penulisan seperti diatas? mungkin _use casenya_ ketika assign nilai yang bergantung kolom tertentu? saya sendiri belum pernah nyoba sih. tapi jika dibaca secara _literally_, sepertinya lebih mudah membaca `dataframe.assign(KOLOM_A=dataframe.KOLOM_A+10)` dibandingkan `dataframe.KOLOM_A = dataframe.KOLOM_A + 10`. dan tentunya bisa dilakukan _chain method_ juga! jadi bisa makin seru. 🤣🤣🤣

tapi rasanya untuk troubleshooting, kayanya lebih asik pakai style ini.

```python
# asumsikan dataframe memiliki kolom tunggal bernama KOLOM_A
dataframe_baru = dataframe.assign(
        KOLOM_B=dataframe.KOLOM_A + 10
    ).assign(
        KOLOM_C=lambda x: x.KOLOM_B * 20,
        KOLOM_D=dataframe.KOLOM_A.diff()
    )
```
